### ISSUE TITLE
Possibly replace usage of System.Numerics with Silk.NET.Maths everywhere

### ISSUE THREAD

**Issue by joffarex** [Ultz/Silk.NET#401]
Mainly interested in this, as I have seen that some modules use new Maths library, but some of them still use default `System.Numerics` one.

I think, just for stability reasons, `Silk.NET.Maths` should be used even in OpenGL bindings, or windowing abstractions.

If this is not *yet* implemented, I could try and integrate it wherever possible.

**Comment by Perksey** [comment]
Not replacing our usage of System.Numerics was an executive decision made in response to consumer feedback gathered in the early days of Silk.NET.

Ideally we'd use System.Numerics everywhere but their APIs just aren't there. However their APIs are much more performant than ours, and even once we've finished adding hardware intrinsics to Silk.NET.Maths, numerics will likely still be faster because it gets special treatment from the JIT.

We primarily created Silk.NET.Maths to patch in "holes" left by System.Numerics. However, we understand some users will want to use one library through and through. I recommend using Numerics if you're only concerned with floats.

If not done so already, our APIs should be updated to accept Silk.NET.Maths types as an alternative, as I'd like the two maths libraries to peacefully coexist. I believe Kai was looking at potentially adding explicit casts to/from Numerics types from/to Silk types (if not done so already) to make this easier.

TL;DR Replace Numerics entirely, we will not, however we'd like to make using Silk maths with the greater library easier.

For further comments @HurricanKai is probably best so assigning to him.

**Comment by joffarex** [comment]
Casts would be wonderful if implemented, so on the client side, one could default to use only one or another. 

Thanks for clear explanation!

**Comment by HurricanKai** [comment]
2.1 or 2.2 will definitely include lots of improvements to maths right now in regards to this issue, I'm looking at replacing the .ToSystem() / .ToGeneric() methods with implicit casts.
I'm also constantly hunting for things were we don't provide options for either our or System maths, so any PRs improving this are welcome.

**Comment by joffarex** [comment]
So to make it clear, you want to expose API which both versions? something like, for example, `Uniform3(Vector3 v)` and overload `Uniform3<T>(Vector3d<T> v)`? If so, I could start working on that in near future.

**Comment by Perksey** [comment]
Yeah preferably. Obviously not necessarily every OpenGL function will support every type Silk.NET.Maths supports, so there may have to be some checking for that, but yeah that'll be a great addition!

**Comment by joffarex** [comment]
Okay then, In few days I will start a pull request with those changes in mind. Hopefully everything will go well, I will close this issue for now.

**Comment by Alan-FGR** [comment]
> 2.1 or 2.2 will definitely include lots of improvements to maths right now in regards to this issue, I'm looking at replacing the .ToSystem() / .ToGeneric() methods with implicit casts. I'm also constantly hunting for things were we don't provide options for either our or System maths, so any PRs improving this are welcome.

I see those methods haven't been replaced yet. Is there a reason these conversions are not reinterpret casts btw? I personally would much prefer a lib to give priority to the System.Numerics types and only fill the holes when necessary and preferably with generic math code since that's possible now.
Anyway, keep up the awesome work ;)
